src/lib/remark-plugins/remark-plugin-anchor-link-data/index.tsx

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import { Plugin, Transformer } from 'unified'
+import { visit, Node } from 'unist-util-visit'
+import { Heading } from 'mdast'
+import toString from 'mdast-util-to-string'
+import guaranteeUniqueSlug from './guarantee-unique-slug'
+
+export type AnchorLinkItem = {
+	id: string
+	level: 1 | 2 | 3 | 4 | 5 | 6
+	title: string
+}
+
+type RemarkPluginAnchorLinkDataOptions = {
+	anchorLinks: AnchorLinkItem[]
+}
+
+/**
+ * A remark plugin that adds `id` and `data-text-content` attributes
+ * to all heading elements in the incoming `mdast` tree.
+ *
+ * It requires an `anchorLinks` array argument, to which it
+ * appends an `{ id, level, title }` object for each heading element
+ * it traverses.
+ *
+ * Note: this is a revised version of our older `anchor-links` from
+ * `@hashicorp/remark-plugins`. This revised version avoids injecting elements,
+ * instead only adding the necessary data for us to render anchor links using
+ * MDX components.
+ */
+const remarkPluginAnchorLinkData: Plugin<
```

Comment:
Again, going heavy on the comments here. This relates closely to 🎟 [Write RFC for refactoring anchor-links plugin's anchor element injection](https://app.asana.com/0/1100423001970639/1202560954440296/f) - while this implementation serves the scope of this PR for now, I think it could be useful in helping us meet our refactoring goals mentioned in the task.